### ISSUE TITLE
ArgumentTypeCoercion false positives on custom `EloquentBuilder` subclasses after widening `where` closure to `:mixed`

### ISSUE THREAD

**Issue by alies-dev** [psalm/laravel-psalm-plugin#815]
## Summary

PR #784 (commit 9617018, "Fix `InvalidArgument` on arrow-function closures in `Builder::where` family") widened the `Builder::where` closure param return type from `Closure(Builder):Builder | Closure(Builder):void` to `Closure(Builder):mixed`.

This successfully reduced `InvalidArgument` on arrow-function closures, but introduced new `ArgumentTypeCoercion` false positives on projects that subclass `Illuminate\Database\Eloquent\Builder` and override `where` with a narrower closure signature.

## Reproduction

Any custom Eloquent builder override hits it, e.g.

```php
final class FooEloquentBuilder extends Builder
{
    public function publishedOnly(): self
    {
        return $this->where(static fn (self $q): self => $q->whereNotNull('published_at'));
    }
}
```

With v4.9.0 the call above reports:

```
ArgumentTypeCoercion: Argument 1 of FooEloquentBuilder::where
 expects ...impure-Closure(Builder<Foo>):mixed | string,
 but parent type impure-Closure(FooEloquentBuilder):void provided
```

The closure satisfies the old union (`Closure(Builder):Builder | Closure(Builder):void`) but not the widened `:mixed` form, because the subclass's own `where` override declares a narrower closure param than the parent stub now advertises.

## Evidence

Observed on a large private Laravel codebase: v4.8.4 → v4.9.0 produces 5 new `ArgumentTypeCoercion` hits, all on custom `EloquentBuilder` subclasses with narrower `where` overrides. 8 additional pre-existing issues shifted message wording at the same lines (cosmetic, due to the widened parent signature).

## Suggested fix

Narrow the widened return union back to something like `Builder|static|void|mixed` rather than just `mixed`, or covariant-type the closure return on the concrete static builder type so subclasses with narrower overrides don't trip coercion checks.

Related: #776, #784.

**Comment by alies-dev** [comment]
### Status: confirmed, blocked on upstream Psalm

Still reproduces on `master` (Psalm 7.0.0-beta19). Retagged `requires-psalm-changes` + `blocked`.

#### Why it isn't a clean false positive

The `where` family stub types the closure parameter as `Closure(self<TModel>): mixed`, i.e. `Closure(Builder<Foo>)`. A custom builder passes a closure whose parameter is the *narrower* subclass (`FooEloquentBuilder`). By contravariance, a `Closure(FooEloquentBuilder)` is **not** a valid `Closure(Builder<Foo>)`, so Psalm's `ArgumentTypeCoercion` is statically sound. It is only runtime-safe because Laravel always invokes the callback with `$this` (a subclass instance).

#### The correct fix, and why it can't land yet

The accurate annotation is `Closure(static)` (this is what Larastan ships: `(\Closure(static): void)|(\Closure(static): static)`). PHPStan substitutes `static` in a closure-**parameter** position against the call receiver. Psalm 7 does not: it binds `static` to the **enclosing analyzed class**, breaking the common untyped/base-`Builder` closure case.

Minimal vanilla repro (no plugin):

```php
/** @template T */
class Box {
    /** @param (\Closure(static): mixed) $cb */
    public function run($cb): void {}
}
class Caller {
    public function c(Box $box): void {
        $box->run(fn ($q) => $q->run(fn ($x) => $x));
    }
}
```

```
InvalidArgument: Argument 1 of Box::run expects impure-Closure(Box&static):mixed, but pure-Closure(static):mixed provided
UndefinedMethod: Method Caller::run does not exist
```

`static` resolved to `Caller` (the enclosing class) instead of the receiver `Box`. Switching our stub to `static` therefore trades this false positive for a worse regression on every untyped/base-`Builder` closure.

| stub | custom-builder narrower closure | untyped / base-`Builder` closure |
|---|---|---|
| `self<TModel>` (current) | ❌ `ArgumentTypeCoercion` | ✅ clean |
| `static` (Larastan's form) | ✅ | ❌ `static` binds to enclosing class |

#### Upstream tracking

No vimeo/psalm issue covers `static`-in-closure-parameter substitution exactly. Closest umbrella issues for the resolution gap:

- vimeo/psalm#7244 — templates/`static` not resolved in closures/callables when invoked.
- vimeo/psalm#5820 — closure parameters analysed without the function (generic) context.

Sibling, already-known limitation around `$this`/`static` in generic type parameters: vimeo/psalm#11770 (and merged #11768). PHPStan-parity tag that several frameworks ship but Psalm lacks: vimeo/psalm#11851 / #11853 (`@param-closure-this`).

We flip `self<TModel>` → `static` in the `where` family once Psalm resolves `static` in closure-parameter position against the receiver.

#### Workaround (works today)

Widen the closure parameter to `Builder`, or leave it untyped:

```php
return $this->where(fn (Builder $q) => $q->whereNotNull('published_at'));
```

**Comment by alies-dev** [comment]
Converted to the repository internal ideas list (`.alias/issues.md`) so GitHub issues stay focused on active work. Reopen or recreate this when it is prioritized.
